File: sendgift.py
"""
sendgift.py — Telegram account session orqali gift yuboruvchi server
Bot.js dan HTTP so'rov oladi va Telethon bilan gift yuboradi.

O'rnatish:
  pip install telethon python-dotenv aiohttp

Ishlatish:
  python sendgift.py
"""

import asyncio
import os
import json
import logging
from aiohttp import web
from telethon import TelegramClient
from telethon.tl.functions.payments import SendStarsFormRequest, GetPaymentFormRequest
from telethon.tl.functions.messages import SendMediaRequest
from telethon.tl.types import InputInvoiceStarGift, InputInvoiceSlug
from telethon.errors import SessionPasswordNeededError
from dotenv import load_dotenv

load_dotenv()

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s'
)
log = logging.getLogger(__name__)

API_ID   = int(os.getenv('API_ID', '0'))
API_HASH = os.getenv('API_HASH', '')
PHONE    = os.getenv('PHONE', '')
SESSION  = os.getenv('SESSION_NAME', 'gift_sender')
HOST     = os.getenv('SENDGIFT_HOST', '127.0.0.1')
PORT     = int(os.getenv('SENDGIFT_PORT', '4242'))
SECRET   = os.getenv('SENDGIFT_SECRET', 'supersecretkey')

client: TelegramClient = None


async def ensure_client():
    global client
    if client and client.is_connected():
        return client
    client = TelegramClient(SESSION, API_ID, API_HASH)
    await client.connect()
    if not await client.is_user_authorized():
        log.warning("Session mavjud emas. Yangi login kerak!")
        await client.send_code_request(PHONE)
        code = input("Telegram kodni kiriting: ")
        try:
            await client.sign_in(PHONE, code)
        except SessionPasswordNeededError:
            pw = input("2FA parolini kiriting: ")
            await client.sign_in(password=pw)
        log.info("Login muvaffaqiyatli!")
    return client


async def handle_send_gift(request: web.Request) -> web.Response:
    """
    POST /sendgift
    Body JSON:
    {
      "secret": "...",
      "gift_id": "telegram_gift_id_or_slug",
      "user_id": 123456789,
      "anonymous": false,
      "sender_name": "Botdan sovg'a"
    }
    """
    try:
        auth = request.headers.get('X-Secret', '')
        if auth != SECRET:
            return web.json_response({'ok': False, 'error': 'Unauthorized'}, status=401)

        data = await request.json()
        gift_id     = str(data.get('gift_id', ''))
        user_id     = int(data.get('user_id', 0))
        anonymous   = bool(data.get('anonymous', False))
        sender_name = str(data.get('sender_name', ''))

        if not gift_id or not user_id:
            return web.json_response({'ok': False, 'error': 'gift_id va user_id kerak'}, status=400)

        tg = await ensure_client()

        # Foydalanuvchi entitisini olish
        try:
            peer = await tg.get_entity(user_id)
        except Exception as e:
            log.error(f"Foydalanuvchi topilmadi: {user_id} — {e}")
            return web.json_response({'ok': False, 'error': f'User topilmadi: {e}'}, status=404)

        # Gift yuborish (Telethon InputInvoiceStarGift)
        try:
            invoice = InputInvoiceStarGift(
                peer=peer,
                gift_id=int(gift_id),
                hide_name=anonymous,
                include_upgrade=False,
                message=None
            )
            form = await tg(GetPaymentFormRequest(invoice=invoice))
            result = await tg(SendStarsFormRequest(
                form_id=form.form_id,
                invoice=invoice
            ))
            log.info(f"Gift yuborildi: {gift_id} → {user_id}")
            return web.json_response({'ok': True, 'result': str(result)})
        except Exception as e:
            log.error(f"Gift yuborishda xato: {e}")
            return web.json_response({'ok': False, 'error': str(e)}, status=500)

    except Exception as e:
        log.error(f"Handle xatosi: {e}")
        return web.json_response({'ok': False, 'error': str(e)}, status=500)


async def handle_health(request: web.Request) -> web.Response:
    tg = await ensure_client()
    me = await tg.get_me()
    return web.json_response({
        'ok': True,
        'account': f"@{me.username}",
        'name': f"{me.first_name or ''} {me.last_name or ''}".strip()
    })


async def main():
    log.info(f"SendGift server ishga tushmoqda: {HOST}:{PORT}")
    await ensure_client()
    me = await client.get_me()
    log.info(f"Hisob: @{me.username} | {me.first_name}")

    app = web.Application()
    app.router.add_post('/sendgift', handle_send_gift)
    app.router.add_get('/health', handle_health)

    runner = web.AppRunner(app)
    await runner.setup()
    site = web.TCPSite(runner, HOST, PORT)
    await site.start()
    log.info(f"✅ SendGift server tayyor: http://{HOST}:{PORT}")
    await asyncio.Event().wait()


if __name__ == '__main__':
    asyncio.run(main())

File: sendgift.js
require('dotenv').config();
const express = require('express');
const fs = require('fs').promises;
const path = require('path');
const { TelegramClient, Api } = require('telegram');
const { StringSession } = require('telegram/sessions');

const API_ID = Number(process.env.API_ID || 0);
const API_HASH = process.env.API_HASH || '';
const SESSION_FILE = path.join(__dirname, 'telegram.session');
const HOST = process.env.SENDGIFT_HOST || '0.0.0.0';
const PORT = Number(process.env.SENDGIFT_PORT || 4242);
const SECRET = process.env.SENDGIFT_SECRET || 'supersecretkey';

let client = null;
let authState = null;

function deferred() {
  let resolve, reject;
  const promise = new Promise((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return { promise, resolve, reject };
}

async function loadSessionString() {
  try {
    return (await fs.readFile(SESSION_FILE, 'utf8')).trim();
  } catch {
    return '';
  }
}

async function saveSessionString(sessionString) {
  await fs.writeFile(SESSION_FILE, sessionString, 'utf8');
}

async function getClient() {
  if (client) return client;
  if (!API_ID || !API_HASH) throw new Error('API_ID yoki API_HASH yo‘q');
  const sessionString = await loadSessionString();
  client = new TelegramClient(new StringSession(sessionString), API_ID, API_HASH, {
    connectionRetries: 5
  });
  await client.connect();
  return client;
}

async function isAuthorized() {
  const c = await getClient();
  return await c.isUserAuthorized();
}

function requireSecret(req, res, next) {
  if ((req.headers['x-secret'] || '') !== SECRET) {
    return res.status(401).json({ ok: false, error: 'Unauthorized' });
  }
  next();
}

async function startLogin(phone) {
  const c = await getClient();

  if (authState?.running) {
    throw new Error('Login jarayoni allaqachon davom etmoqda');
  }

  const codeGate = deferred();
  const passGate = deferred();
  const doneGate = deferred();
  const passwordNeededGate = deferred();

  authState = {
    running: true,
    phone,
    codeGate,
    passGate,
    doneGate,
    passwordNeededGate,
    needPassword: false,
    lastError: null
  };

  (async () => {
    try {
      await c.start({
        phoneNumber: async () => phone,
        phoneCode: async () => {
          authState.needCode = true;
          return await codeGate.promise;
        },
        password: async () => {
          authState.needPassword = true;
          passwordNeededGate.resolve(true);
          return await passGate.promise;
        },
        onError: (err) => {
          authState.lastError = err?.message || String(err);
        }
      });

      const sessionString = c.session.save();
      await saveSessionString(sessionString);
      const me = await c.getMe();
      doneGate.resolve({
        ok: true,
        user: {
          id: me?.id ? String(me.id) : null,
          username: me?.username || null,
          first_name: me?.firstName || me?.first_name || null,
          last_name: me?.lastName || me?.last_name || null
        }
      });
    } catch (err) {
      doneGate.reject(err);
    } finally {
      authState.running = false;
    }
  })().catch(err => {
    authState.lastError = err?.message || String(err);
    doneGate.reject(err);
    authState.running = false;
  });

  return { ok: true, need_code: true };
}

async function submitCode(code) {
  if (!authState?.running) return { ok: false, error: 'Login boshlanmagan' };
  authState.codeGate.resolve(code);
  const race = await Promise.race([
    authState.doneGate.promise.then(v => ({ type: 'done', value: v })).catch(err => ({ type: 'done_err', error: err })),
    authState.passwordNeededGate.promise.then(() => ({ type: 'need_password' })),
    new Promise(resolve => setTimeout(() => resolve({ type: 'pending' }), 1200))
  ]);

  if (race.type === 'done') return race.value;
  if (race.type === 'done_err') return { ok: false, error: race.error?.message || String(race.error) };
  if (race.type === 'need_password') return { ok: false, need_password: true };
  return { ok: true, pending: true };
}

async function submitPassword(password) {
  if (!authState?.running) return { ok: false, error: 'Login boshlanmagan' };
  authState.passGate.resolve(password);
  try {
    const result = await authState.doneGate.promise;
    return result;
  } catch (err) {
    return { ok: false, error: err?.message || String(err) };
  }
}

async function resolveRecipient(c, recipient) {
  const value = String(recipient || '').trim();
  if (!value) throw new Error('recipient kerak');

  if (value === 'me') {
    return await c.getInputEntity('me');
  }

  if (/^@?[a-zA-Z0-9_]{5,32}$/.test(value)) {
    const username = value.startsWith('@') ? value : `@${value}`;
    return await c.getInputEntity(username);
  }

  if (/^\d+$/.test(value)) {
    return await c.getInputEntity(Number(value));
  }

  return await c.getInputEntity(value);
}

async function sendGiftToRecipient({ recipient, giftId, anonymous = false, message = '' }) {
  const c = await getClient();
  if (!(await isAuthorized())) {
    throw new Error('Telegram akkaunt ulanmagan');
  }

  const peer = await resolveRecipient(c, recipient);
  const invoice = new Api.InputInvoiceStarGift({
    peer,
    giftId: BigInt(giftId),
    hideName: !!anonymous,
    includeUpgrade: false,
    message: message
      ? new Api.TextWithEntities({
          text: String(message),
          entities: []
        })
      : undefined
  });

  const form = await c.invoke(new Api.payments.GetPaymentForm({ invoice }));
  const result = await c.invoke(new Api.payments.SendStarsForm({
    formId: form.formId,
    invoice
  }));

  return { ok: true, result: String(result) };
}

async function health() {
  try {
    const authorized = await isAuthorized();
    return { ok: true, authorized, pending: !!authState?.running };
  } catch (err) {
    return { ok: false, error: err?.message || String(err) };
  }
}

async function main() {
  const app = express();
  app.use(express.json({ limit: '1mb' }));

  app.get('/health', requireSecret, async (_req, res) => {
    res.json(await health());
  });

  app.post('/auth/start', requireSecret, async (req, res) => {
    try {
      const phone = String(req.body?.phone || '').trim();
      if (!phone) return res.status(400).json({ ok: false, error: 'phone kerak' });
      res.json(await startLogin(phone));
    } catch (err) {
      res.status(500).json({ ok: false, error: err?.message || String(err) });
    }
  });

  app.post('/auth/code', requireSecret, async (req, res) => {
    try {
      const code = String(req.body?.code || '').trim();
      if (!code) return res.status(400).json({ ok: false, error: 'code kerak' });
      res.json(await submitCode(code));
    } catch (err) {
      res.status(500).json({ ok: false, error: err?.message || String(err) });
    }
  });

  app.post('/auth/password', requireSecret, async (req, res) => {
    try {
      const password = String(req.body?.password || '');
      if (!password) return res.status(400).json({ ok: false, error: 'password kerak' });
      res.json(await submitPassword(password));
    } catch (err) {
      res.status(500).json({ ok: false, error: err?.message || String(err) });
    }
  });

  app.post('/sendgift', requireSecret, async (req, res) => {
    try {
      const recipient = req.body?.recipient;
      const giftId = req.body?.gift_id;
      const anonymous = !!req.body?.anonymous;
      const message = String(req.body?.message || '');

      if (!recipient) return res.status(400).json({ ok: false, error: 'recipient kerak' });
      if (!giftId) return res.status(400).json({ ok: false, error: 'gift_id kerak' });

      const result = await sendGiftToRecipient({ recipient, giftId, anonymous, message });
      res.json(result);
    } catch (err) {
      res.status(500).json({ ok: false, error: err?.message || String(err) });
    }
  });

  await getClient();
  app.listen(PORT, HOST, () => {
    console.log(`✅ sendgift.js ishga tushdi: http://${HOST}:${PORT}`);
  });
}

main().catch(err => {
  console.error('sendgift main error:', err);
  process.exit(1);
});

process.once('SIGINT', () => {
  if (client) client.destroy();
  process.exit(0);
});
process.once('SIGTERM', () => {
  if (client) client.destroy();
  process.exit(0);
});
